File: solutions/0039. Combination Sum/0039.py
# Time complexity: O(n^t)
# Space complexity: O(t)
class Solution:
  def combinationSum(
    self,
    candidates: List[int],
    target: int
  ) -> List[List[int]]:
    ans = []

    def dfs(s: int, target: int, path: List[int]) -> None:
      if target < 0:
        return
      if target == 0:
        ans.append(path.copy())
        return

      for i in range(s, len(candidates)):
        path.append(candidates[i])
        dfs(i, target - candidates[i], path)
        path.pop()

    candidates.sort()
    dfs(0, target, [])

    return ans

File: solutions/0039. Combination Sum/0039.rs
// Time complexity: O(n^t)
// Space complexity: O(t)
impl Solution {
    pub fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut ans = Vec::new();
        candidates.sort_unstable();

        fn dfs(
            candidates: &Vec<i32>,
            start: usize,
            target: i32,
            path: &mut Vec<i32>,
            ans: &mut Vec<Vec<i32>>,
        ) {
            if target < 0 {
                return;
            }
            if target == 0 {
                ans.push(path.clone());
                return;
            }

            for i in start..candidates.len() {
                path.push(candidates[i]);
                dfs(candidates, i, target - candidates[i], path, ans);
                path.pop();
            }
        }

        let mut path = Vec::new();
        dfs(&candidates, 0, target, &mut path, &mut ans);

        ans
    }
}
